test/CoinSelection/UtxoIndex.purs

```diff
@@ -0,0 +1,168 @@
+
+suite :: TestPlanM (Aff Unit) Unit
+suite =
+  group "UtxoIndex" do
+    test "prop_buildUtxoIndex_empty" do
+      UtxoIndex.buildUtxoIndex Map.empty `shouldEqual` UtxoIndex.emptyUtxoIndex
+
+    test "prop_buildUtxoIndex_invariant" do
+      quickCheck prop_buildUtxoIndex_invariant
+
+    test "prop_utxoIndexInsertEntry_invariant" do
+      quickCheck prop_utxoIndexInsertEntry_invariant
+
+    test "prop_utxoIndexDeleteEntry_invariant" do
+      quickCheck prop_utxoIndexDeleteEntry_invariant
+
+prop_buildUtxoIndex_invariant :: ArbitraryUtxoMap -> QuickCheck.Result
+prop_buildUtxoIndex_invariant =
+  invariantHolds <<< UtxoIndex.buildUtxoIndex <<< unwrap
+
+prop_utxoIndexInsertEntry_invariant
+  :: ArbitraryTxUnspentOut -> ArbitraryUtxoIndex -> QuickCheck.Result
+prop_utxoIndexInsertEntry_invariant entry utxoIndex =
+  invariantHolds $
+    UtxoIndex.utxoIndexInsertEntry (unwrap entry) (unwrap utxoIndex)
+
+prop_utxoIndexDeleteEntry_invariant
+  :: ArbitraryTxUnspentOut -> ArbitraryUtxoIndex -> QuickCheck.Result
+prop_utxoIndexDeleteEntry_invariant entry utxoIndex =
+  invariantHolds $
+    UtxoIndex.utxoIndexDeleteEntry (unwrap entry) (unwrap utxoIndex)
+
+-- | Taken from cardano-wallet:
+-- | https://github.com/input-output-hk/cardano-wallet/blob/9d73b57e23392e25148cfc8db560cb8f656cb56a/lib/primitive/test/spec/Cardano/Wallet/Primitive/Types/UTxOIndexSpec.hs#L183
+invariantHolds :: UtxoIndex -> QuickCheck.Result
+invariantHolds utxoIndex =
+  case UtxoIndex.checkUtxoIndexInvariants utxoIndex of
+    InvariantHolds -> QuickCheck.Success
+    status -> QuickCheck.Failed (show status)
+
+--------------------------------------------------------------------------------
+-- Arbitrary
+--------------------------------------------------------------------------------
+
+newtype ArbitraryUtxoIndex = ArbitraryUtxoIndex UtxoIndex
+
+derive instance Newtype ArbitraryUtxoIndex _
+
+instance Arbitrary ArbitraryUtxoIndex where
+  arbitrary =
+    (arbitrary :: Gen ArbitraryUtxoMap)
+      <#> wrap <<< UtxoIndex.buildUtxoIndex <<< unwrap
+
+newtype ArbitraryUtxoMap = ArbitraryUtxoMap UtxoMap
+
+derive instance Generic ArbitraryUtxoMap _
+derive instance Newtype ArbitraryUtxoMap _
+
+instance Show ArbitraryUtxoMap where
+  show = genericShow
+
+instance Arbitrary ArbitraryUtxoMap where
+  arbitrary = wrap <$> Map.genMap genTransactionInput genTransactionOutput
+
+newtype ArbitraryTxUnspentOut =
+  ArbitraryTxUnspentOut (TransactionInput /\ TransactionOutput)
+
+derive instance Newtype ArbitraryTxUnspentOut _
+
+instance Arbitrary ArbitraryTxUnspentOut where
+  arbitrary = wrap <$> lift2 Tuple genTransactionInput genTransactionOutput
+
+genTransactionInput :: Gen TransactionInput
+genTransactionInput = unwrap <$> (arbitrary :: Gen ArbitraryTransactionInput)
+
+genTransactionOutput :: Gen TransactionOutput
+genTransactionOutput = unwrap <$> (arbitrary :: Gen ArbitraryTransactionOutput)
+
+newtype ArbitraryTransactionInput =
```

Comment:
Because there can be multiple ways to write this instance for `TransactionInput` depending on requirements, e.g. the way output indices should be generated may be different, etc.